input-profile.ts

```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    getGamepadObjectsFromProfile(profile: any, obj: Object3D) {
+        const components = profile.layouts[this._handedness].components;
+        if (!components) return;
+
+        for (const i in components) {
+            const visualResponses = components[i].visualResponses;
+
+            for (const j in visualResponses) {
+                // update buttons with new interface of current visual response
+                const visualResponse = visualResponses[j];
+                const valueNode = visualResponse.valueNodeName;
+                const minNode = visualResponse.minNodeName;
+                const maxNode = visualResponse.maxNodeName;
+
+                this.getGamepadObjectByName(obj, valueNode);
+                this.getGamepadObjectByName(obj, minNode);
+                this.getGamepadObjectByName(obj, maxNode);
+
+                let indice = visualResponses[j].componentProperty;
+                const response: VisualResponse = {
+                    target: this._gamepadObjects[valueNode],
+                    min: this._gamepadObjects[minNode],
+                    max: this._gamepadObjects[maxNode],
+                    id: components[i].gamepadIndices[indice], // Assign a unique ID
+                };
+                switch (indice) {
+                    case 'button':
+                        this._buttons.push(response);
+                        break;
+                    case 'xAxis':
+                        this._axes.push(response);
+                        break;
+                    case 'yAxis':
+                        this._axes.push(response);
+                        break;
+                }
+            }
+        }
+    }
+
+    getGamepadObjectByName(obj: Object3D, name: string) {
+        if (!obj || !name) return;
+
+        if (obj.name === name) this._gamepadObjects[name] = obj;
+
+        const children = obj.children;
+        for (let i = 0; i < children.length; ++i)
+            this.getGamepadObjectByName(children[i], name);
+    }
+
+    assignTransform(target: Object3D, min: Object3D, max: Object3D, value: number) {
+        vec3.lerp(
+            _TempVec,
+            min.getPositionWorld(minTemp),
+            max.getPositionWorld(maxTemp),
+            value
+        );
+        target.setPositionWorld(_TempVec);
+        quat.lerp(
+            _TempQuat,
+            min.getRotationWorld(_TempRotation1),
+            max.getRotationWorld(_TempRotation2),
+            value
+        );
+        quat.normalize(_TempQuat, _TempQuat);
+        target.setRotationWorld(_TempQuat);
+    }
+
+    mapGamepadInput() {
+        for (const button of this._buttons) {
+            const ButtonValue = this._gamepad?.buttons[button.id].value || 0;
+            this.assignTransform(button.target, button.min, button.max, ButtonValue);
+        }
+        for (const axis of this._axes) {
+            const AxisValue = this._gamepad?.axes[axis.id] || 0;
```

Comment:
Gamepad can't be `null` here, please use `!` instead of `?`

Reply:
Also, do not use `|| 0`. You sjould only push buttons that have a correct mapping in gamepad.